core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaSchemaRegistry.java

```diff
@@ -344,16 +345,21 @@ public int register(String subject,
       }
 
       // determine the latest version of the schema in the subject
-      Iterator<Schema> allVersions = getAllVersions(subject);
+      Iterator<Schema> allVersions = getAllVersions(subject, true);
+      Iterator<Schema> undeletedVersions = getAllVersions(subject, false);
+
       List<String> allSchemas = new ArrayList<>();
       Schema latestSchema = null;
       int newVersion = MIN_VERSION;
       while (allVersions.hasNext()) {
-        latestSchema = allVersions.next();
+        newVersion = allVersions.next().getVersion() + 1;
+      }
+      while (undeletedVersions.hasNext()) {
+        latestSchema = undeletedVersions.next();
```

Comment:
Would it be worth re-naming `allSchemas`? It isn't actually `allSchemas` anymore -- it's really `undeletedVersions`. In fact, aside from the fact that `undeletedVersions` is currently an iterator, I would suggest using `undeletedVersions` instead.